### PR TITLE
Convert ha-card to LitElement and TypeScript

### PR DESCRIPTION
**Pull request by thomasloven**

### CODE REVIEW

File: src/components/ha-card.ts [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+  TemplateResult,
+  PropertyDeclarations,
+  LitElement,
+} from "lit-element";
+
+class HaCard extends LitElement {
+  protected render(): TemplateResult | void {
+    return html`
+      <style include="paper-material-styles">
+        :host {
```
Comment (balloob) [inline src/components/ha-card.ts]: Do the CSS like this: https://github.com/home-assistant/home-assistant-polymer/blob/7a16961387e1be5e0b2826c722d732b09548dbd0/src/panels/lovelace/components/hui-generic-entity-row.ts#L88-L90

File: src/components/ha-card.ts [hunk 2]
```diff
@@ -0,0 +1,48 @@
...
+          letter-spacing var(--paper-font-headline_-_letter-spacing);
+          line-height var(--paper-font-headline_-_line-height);
+          -webkit-font-smoothing: var(--paper-font-headline_-_-webkit-font-smoothing);
+          text-rendering: var(--paper-font-common-expensive-kerning_-_text-rendering);
+          opacity: var(--dark-primary-opacity);
+          padding: 24px 16px 16px;
+        }
+      </style>
+
+      <template is="dom-if" if="${this.header}">
```
Comment (balloob) [inline src/components/ha-card.ts]: `dom-if` is a Polymer thing, use turnery statements

File: src/components/ha-card.ts [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+        <div class="header">${this.header}</div>
+      </template>
+      <slot></slot>
+    `;
+  }
+
+  static get properties(): PropertyDeclarations {
+    return {
+      header: String,
```
Comment (balloob) [inline src/components/ha-card.ts]:
```suggestion
      header: {},
```

File: src/components/ha-card.ts [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+  css,
+  CSSResult,
+  html,
+  LitElement,
+  property,
+  TemplateResult,
+} from "lit-element";
+
+class HaCard extends LitElement {
+  @property() public header?: {};
```
Comment (balloob) [inline src/components/ha-card.ts]:
```suggestion
  @property() public header?: string;
```

File: src/components/ha-card.ts [hunk 5]
```diff
@@ -0,0 +1,49 @@
...
+          var(--paper-card-background-color, white)
+        );
+        border-radius: var(--ha-card-border-radius, 2px);
+        box-shadow: 0 2px 2px 0 rgba(0, 0, 0, 0.14),
+          0 1px 5px 0 rgba(0, 0, 0, 0.12), 0 3px 1px -2px rgba(0, 0, 0, 0.2);
+        color: var(--primary-text-color);
+        display: block;
+        transition: all 0.3s ease-out;
+      }
+      .header:not(:empty) {
```
Comment (balloob) [inline src/components/ha-card.ts]: I did not know this was a thing.

File: src/components/ha-card.ts [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+        -webkit-font-smoothing: antialiased;
+        font-family: Roboto, Noto, sans-serif;
```
Comment (balloob) [inline src/components/ha-card.ts]: I don't think this is necessary, already defined on the body?

Comment (balloob) [reply src/components/ha-card.ts]: (bunch of these other props too)

File: src/components/ha-card.ts (hunk 4 above)
Comment (thomasloven) [reply src/components/ha-card.ts]: Right. Leftover from when I still tried to get more advanced headers to work... like litHtml templates with toggles.

File: src/components/ha-card.ts (hunk 5 above)
Comment (thomasloven) [reply src/components/ha-card.ts]: Really nice function.

I tried to put the header in a named `slot`, but that didn't work with the styling when empty. So I found the suggestion `.header:not(:has(slot:has-assigned))`, which I rejected for maintainability reasons... and for the fact that `:has-assigned` isn't specified yet.

File: src/components/ha-card.ts (hunk 6 above)
Comment (thomasloven) [reply src/components/ha-card.ts]: Turns out `text-rendering: optimizeLegibility` only matters for font size < 20 px anyway...

File: src/components/ha-card.ts (hunk 4 above)
Comment (balloob) [reply src/components/ha-card.ts]: `string | TemplateResult` would be the correct type then 👍

Comment (balloob) [reply src/components/ha-card.ts]: However, that should be done using named slots.

**Comment by balloob** [comment]
Nice !

**Pull request merged**